src/Parsers/ExpressionElementParsers.cpp

```diff
@@ -972,8 +973,56 @@ bool ParserStringLiteral::parseImpl(Pos & pos, ASTPtr & node, Expected & expecte
 
     String s;
 
+    auto makeStringLiteral = [&](String str) -> bool
+    {
+        auto literal = std::make_shared<ASTLiteral>(str);
+        literal->begin = pos;
+        literal->end = ++pos;
+        node = literal;
+        return true;
+    };
+
     if (pos->type == TokenType::StringLiteral)
     {
+        const char * token_begin = pos->begin;
+        const char * token_end = pos->end;
+
+        auto makeHexOrBinStringLiteral = [&] (bool hex, size_t word_size) -> bool
+        {
+            const char * str_begin = token_begin + 2;
+            const char * str_end = token_end - 1;
+            if (str_begin == str_end)
+                return makeStringLiteral("");
+
+            PODArray<UInt8> res;
+            res.resize(pos->size() / word_size + 1);
+            char * res_begin = reinterpret_cast<char *>(res.data());
+            char * res_pos = res_begin;
+            if (hex)
+            {
+                hexStringDecode(str_begin, str_end, res_pos);
+            }
+            else
+            {
+                binStringDecode(str_begin, str_end, res_pos);
+            }
+
+            res.resize(res_pos - res_begin);
+            return makeStringLiteral(String(reinterpret_cast<char *>(res.data())));
```

Comment:
I'd specify the size in the string constructor as well. The result of binStringDecode/hexStringDecode is null terminated (and resizing in the line above is not required, I suppose), but just for safety and to be more explicit.